### ISSUE TITLE
Semantic highlighting for Python: Pylance issue 2023

### ISSUE THREAD

**Issue by finnlogin** [VSCodium/vscodium#1556]
Hi!
I have a problem with semantic highlighting in VSCodium (screenshot: VSCodium is on the left, VSCode is on the right). I have the same issue on MacOS and Linux Mint. After research, I found out that it has to do with the fact that Microsoft prevents PyLance language server from running in unofficial builds. In this [thread](https://github.com/VSCodium/vscodium/pull/935) there is a solution; however, it seems outdated because it doesn't work anymore.

Has anyone encountered this issue and managed to find a solution?

![EArlF](https://github.com/VSCodium/vscodium/assets/69312175/0212e3e9-f70d-4d84-a8a4-2b63e1f3d5e1)

**Comment by EndlessFractal** [comment]
Link to the last working version:

> https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.6.40/vspackage

Works perfectly as of writing.

**Comment by EndlessFractal** [comment]
Link to the last working version:

> https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.6.40/vspackage

Works perfectly as of writing.

**Comment by EndlessFractal** [comment]
Link to the last working version:

> https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.6.40/vspackage

Works perfectly as of writing. Just install it manually by going to extensions, click on the 3 dots and click on **Install from VSIX**, and select the downloaded file.

**Comment by finnlogin** [comment]
It works for me too!
You're my savior! Thanks a bunch!

**Comment by ovo-Tim** [comment]
Do we have any other options？We can't use old version for ever. :cry:

**Comment by finnlogin** [comment]
> 

You can replace the version of the extension in the provided link with the latest one.
https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.8.20/vspackage

**Comment by EndlessFractal** [comment]
> > 
> 
> You can replace the version of the extension in the provided link with the latest one. https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.8.20/vspackage

I've tested the 5 most recent versions (2023.8.21, 2023.8.20, 2023.8.11, 2023.8.10, 2023.7.43) and none work.

Until someone finds a workaround, the last known working version (2023.6.40) will have to do.

A screenshot of the latest VSCodium running the latest Pylance, notice that the color highlighting does not work.
![image](https://github.com/VSCodium/vscodium/assets/89462237/e1167b7c-229b-4d4e-82c1-153d6056c625)

Cheers!

**Comment by finnlogin** [comment]
> > > 
> > 
> > 
> > You can replace the version of the extension in the provided link with the latest one. https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.8.20/vspackage
> 
> I've tested the 5 most recent versions (2023.8.21, 2023.8.20, 2023.8.11, 2023.8.10, 2023.7.43) and none work.
> 
> Until someone finds a workaround, the last known working version (2023.6.40) will have to do.
> 
> A screenshot of the latest VSCodium running the latest Pylance, notice that the color highlighting does not work. ![image](https://user-images.githubusercontent.com/89462237/260860658-e1167b7c-229b-4d4e-82c1-153d6056c625.png)
> 
> Edit: @burritosoftware this might be of interest to you.
> 
> Cheers!

You're right! I downloaded and installed the latest version but never tested it. I'm sorry for misleading you guys.

**Comment by awsms** [comment]
I tried everything, assuming it was an issue on my end. :smiling_face_with_tear:
Is it Microsoft enforcing the usage of their own extensions to propritary VSCode-only, like they have done with the Copilot integration, the C# debugger etc?

**Comment by EndlessFractal** [comment]
> Thanks so much for the link @EndlessFractal, I tried everything, assuming it was an issue on my end. 🥲 Do you know if that's Microsoft enforcing the usage of their own extensions to proprietary VSCode-only, like they have done with the Copilot integration, the C# debugger etc?

It is very possible, as they have taken measures in preventing its usage on the later versions, https://github.com/VSCodium/vscodium/pull/935 shows this, as the previous trick of modifying the product.json no longer works.

**Comment by sergei-dyshel** [comment]
Hoping some another `product.json` trick is possible, I tried replacing entire file with the file from MS VScode and Pylance still didn't work 😭  Bye bye semantic highlighting and other goodies....

**Comment by Rantunah** [comment]
I went to check out the console for the `Python Language Server` with the Microsoft extensions and I got this:
![Screenshot_20230905_182955](https://github.com/VSCodium/vscodium/assets/90865007/b0ee59c5-17a9-45bb-a6bf-a59ca04c94fe)

> **2023-09-05 18:24:57.301 [info]** (Client) You may install and use any number of copies of the software only with Microsoft Visual Studio, Visual Studio for Mac, Visual Studio Code, Azure DevOps, Team Foundation Server, and successor Microsoft products and services (collectively, the “Visual Studio Products and Services”) to develop and test your applications. The software is licensed, not sold. This agreement only gives you some rights to use the software. Microsoft reserves all other rights. You may not: work around any technical limitations in the software that only allow you to use it in certain ways; reverse engineer, decompile or disassemble the software, or otherwise attempt to derive the source code for the software, except and to the extent required by third party licensing terms governing use of certain open source components that may be included in the software; remove, minimize, block, or modify any notices of Microsoft or its suppliers in the software; use the software in any way that is against the law or to create or propagate malware; or share, publish, distribute, or lease the software (except for any distributable code, subject to the terms above), provide the software as a stand-alone offering for others to use, or transfer the software or this agreement to any third party.

So I guess that this might be Microsoft's attempt to enforce the license terms and force some users into VSCode.

**Comment by awsms** [comment]
I tried all the builds (2023.6.x up to 2023.8.x), and 2023.8.10 is the last build that works for me. Wondering why it didn't work for you @EndlessFractal

**Comment by EndlessFractal** [comment]
> I tried all the builds (2023.6.x up to 2023.8.x), and 2023.8.10 is the last build that works for me. Wondering why it didn't work for you @EndlessFractal

Have you tried it on a fresh installation?

As you can see below, I'm not able to get it fully working. I have not done any other changes beside installing Codium and the Python and the Pylance extensions.

### Codium 1.82.0 with Pylance v2023.8.10
![image](https://github.com/VSCodium/vscodium/assets/89462237/e2b2c891-1219-4819-8f4b-bb584393f669)

### Codium 1.82.0 with Pylance v2023.6.40
![image](https://github.com/VSCodium/vscodium/assets/89462237/e2673e0f-359e-4d02-994c-3b89cc6d8c6b)

Cheers!

**Comment by felix9743** [comment]
I wrote a solution to run Pylance in its latest version (I postponed as BUG without doing it on purpose): #1640

**Comment by pickx** [comment]
> I wrote a solution to run Pylance in its latest version (I postponed as BUG without doing it on purpose): #1640

This worked for me with `v2023.9.20`. Thank you.

**Comment by mrlee14** [comment]
> I wrote a solution to run Pylance in its latest version (I postponed as BUG without doing it on purpose): #1640

works with v2023.8.20

**Comment by alephpi** [comment]
![image](https://github.com/VSCodium/vscodium/assets/61275421/8e850853-8737-4616-ac33-463e0e2a9186)
Weirdly enough, I cannot even find pylance extension in codium

**Comment by mrlee14** [comment]
> I wrote a solution to run Pylance in its latest version (I postponed as BUG without doing it on purpose): #1640

version 2023.11.10 ,still works!

method:
cd ~/.vscode-oss/extensions/ms-python.vscode-pylance-2023.11.10/dist  
mv extension.bundle.js extension.bundle.js.bak

**Comment by mrlee14** [comment]
> > I wrote a solution to run Pylance in its latest version (I postponed as BUG without doing it on purpose): #1640
> 
> works also with v2023.10.10

version 2023.11.10 ,still works!

method:
cd ~/.vscode-oss/extensions/ms-python.vscode-pylance-2023.11.10/dist
mv extension.bundle.js extension.bundle.js.bak

**Comment by leon-erd** [comment]
> version 2023.11.10 ,still works!

It seems like it does for me as well.
But now my Jupyter Notebooks have problems to connect to the kernel (works if I disable and enable Jupyter extension but fails again on restart of VSCodium). So I will stay on 2023.6.40 for now... :/

**Comment by alecov** [comment]
> Link to the last working version:
> 
> > https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.6.40/vspackage
> 
> Works perfectly as of writing. Just install it manually by going to extensions, click on the 3 dots and click on **Install from VSIX**, and select the downloaded file.

I updated codium and suddenly the extension host stopped working. Extension bisect did not help, but I manually removed this version of pylance and it stopped crashing.

Then, attempting to reinstall pylance from the VSIX package yields "Corrupt ZIP: end of central directory record signature not found".

Has anyone faced this? Any workarounds?

**Comment by EndlessFractal** [comment]
> > Link to the last working version:
> > > https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.6.40/vspackage
> > 
> > 
> > Works perfectly as of writing. Just install it manually by going to extensions, click on the 3 dots and click on **Install from VSIX**, and select the downloaded file.
> 
> I updated codium and suddenly the extension host stopped working. Extension bisect did not help, but I manually removed this version of pylance and it stopped crashing.
> 
> Then, attempting to reinstall pylance from the VSIX package yields "Corrupt ZIP: end of central directory record signature not found".
> 
> Has anyone faced this? Any workarounds?

Sorry @alecov, I can't be of much help. I'm on the latest version and it hasn't crashed a single time. Have you tried reinstalling VSCodium? If it's still giving you issues, you can try this method and install a more recent version. #1641

![image](https://github.com/VSCodium/vscodium/assets/89462237/bdd8a336-7305-44d3-a450-ce7f4d5103a5)

Cheers!

**Comment by felix9743** [comment]
> > Lien vers la dernière version fonctionnelle : 
> > > [https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.6.40/vspackage ](https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-python/vsextensions/vscode-pylance/2023.6.40/vspackage)
> > 
> > 
> > Fonctionne parfaitement au moment de l'écriture.   Installez-le simplement manuellement en allant dans extensions, cliquez sur les 3 points et cliquez sur ** Installer depuis VSIX ** , et sélectionnez le fichier téléchargé.
> 
> J'ai mis à jour Codium et tout à coup, l'hôte de l'extension a cessé de fonctionner. L'extension bisect n'a pas aidé, mais j'ai supprimé manuellement cette version de pylance et elle a cessé de planter.
> 
> Ensuite, en essayant de réinstaller pylance à partir du package VSIX, vous obtenez "ZIP corrompu : signature d'enregistrement de fin de répertoire central introuvable".
> 
> Quelqu'un a-t-il été confronté à cela ? Des solutions de contournement ?

Hi, if that helps you a little, try to remove **Pylance, Python and Jupyter** from your extension list. Restart and close VSCodium and Then check that you no longer have the folders of its extension in your folder `.vscode-oss/extensions` and, **after restarting VSCodium,** check that it is properly updated.

Once all this is done, either you have define the custom value of the `extensionsGallery` in `product.json` and so, you can reinstall the 3 extensions, either you have to download their latest version (but I have not tested like this) and install them. Then you close **VSCodium** and [you make the changes from step 4 (I just explained step 3).](https://github.com/VSCodium/vscodium/discussions/1641)

**After that, normally Pylance and Jupyter (if you use) should work properly.**

**PS:** There was a version or 2 or this trick didn't work anymore but strangely, it's reworking now....

-----

For information, it works for me with the latest versions:
- Windows **10** and VSCodium **v1.84.2**
- Python **v2023.20.0** and Pylance **v2023.11.10**
- Jupyter **v2023.10.11**

**Comment by fithisux** [comment]
Do we need PyLance to work with Python? I do not want Pylance. Are there any other options?

**Comment by EndlessFractal** [comment]
> Do we need PyLance to work with Python? I do not want Pylance. Are there any other options?

You do not. Pylance is just to make the Python code more legible by adding colors, also known as, semantic highlighting.

**Comment by fithisux** [comment]
By the way  @EndlessFractal  there is a set of utils by ms-python for mypy/isort/black/pylint that work very good.

For some cases (maybe I misconfigured) mypy has accurate semantic highlighting if you specify types. I changed a type to str and worked immediately.  What mypy finds, it is reported.

**Comment by alecov** [comment]
@EndlessFractal @felix9743 Thanks mates, the tip from @felix9743 actually worked, although I'm pretty sure I have tried it before posting here.

**Comment by pickx** [comment]
after several tries, confirming that this method does  _not_ work with `2023.12.1` on my machine. 
it works with `2023.11.10`.

**Comment by Splarkszter** [comment]
Confirmed working with `2023.6.40`

Had a bit of a struggle but after disabling and enabling the main Python extension the semantic highlighting is working now!
![image](https://github.com/VSCodium/vscodium/assets/90066319/d84be314-5c50-4958-b80d-bccfa09917e7)
